test/others/criu-ns/run.py

```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python
+
+import multiprocessing
+import os
+import pty
+import shutil
+import subprocess
+import sys
+import time
+
+
+CRIU_BIN = "../../../criu/criu"
+CRIU_NS = "../../../scripts/criu-ns"
+
+
+def check_dumpdir():
+    if os.path.isdir("dumpdir"):
+        shutil.rmtree("dumpdir")
+    os.mkdir("dumpdir", 0o755)
+
+
+def create_pty():
+    fd_m, fd_s = pty.openpty()
+    return (os.fdopen(fd_m, "wb"), os.fdopen(fd_s, "wb"))
+
+
+def test_dump_and_restore_with_shell_job():
+    check_dumpdir()
+    os.setsid()
+
+    with open("running", "w") as file:
+        pass
+    pid = os.fork()
+    if pid == 0:
+        while True:
+            if not os.access("running", os.F_OK):
+                sys.exit(0)
+            time.sleep(1)
+
+    cmd = [CRIU_NS, "dump", "-D", "dumpdir", "-v", "--shell-job",
+           "-t", str(pid), "--criu-binary", CRIU_BIN]
+    ret = subprocess.Popen(cmd).wait()
+    if ret != 0:
+        sys.exit(ret)
+
+    os.unlink("running")
+    fd_m, fd_s = create_pty()
+    pid = os.fork()
+    if pid == 0:
+        fd_m.close()
+        # since criu-ns takes control of the tty stdin
+        os.dup2(fd_s.fileno(), 0)
+        cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v",
+               "--shell-job", "--criu-binary", CRIU_BIN]
+        ret = subprocess.Popen(cmd).wait()
+        if ret != 0:
+            sys.exit(ret)
+        os._exit(0)
+
+    fd_s.close()
+    os.waitpid(pid, 0)
+
+
+def test_dump_and_restore_without_shell_job(restore_detached=False):
+    check_dumpdir()
+
+    with open("running", "w") as file:
+        pass
+    fd_m, fd_s = create_pty()
+    pid = os.fork()
+    fd_m.close()
+    if pid == 0:
+        os.setsid()
+        os.dup2(fd_s.fileno(), 0)
+        os.dup2(fd_s.fileno(), 1)
+        os.dup2(fd_s.fileno(), 2)
+        while True:
+            if not os.access("running", os.F_OK):
+                sys.exit(0)
+            time.sleep(1)
+
+    cmd = [CRIU_NS, "dump", "-D", "dumpdir", "-v", "-t", str(pid),
+           "--criu-binary", CRIU_BIN]
+    ret = subprocess.Popen(cmd).wait()
+    if ret != 0:
+        sys.exit(ret)
+
+    os.unlink("running")
+    fd_m, fd_s = create_pty()
+    pid = os.fork()
+    if pid == 0:
+        if restore_detached:
+            cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v",
+                   "--restore-detached", "--criu-binary",
+                   CRIU_BIN]
+        else:
+            cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v",
+                   "--criu-binary", CRIU_BIN]
+
+        ret = subprocess.Popen(cmd, start_new_session=True).wait()
+        if ret != 0:
+            sys.exit(ret)
```

Comment:
It might be good to add a function that reads the content of `dump.log`/`restore.log` and writes to STDOUT. The call this function in the case of an error.

For example:
```Python
def read_log_file(filename):
    file_path = os.path.join('dumpdir', filename)
    with open(file_path, 'r') as file:
        content = file.read()
        print(content)
```
```Python
read_log_file('restore.log')
```

Reply:
I really prefer the current approach that CI logs (or `stdout` output) are always visible regardless of there is an error or not :)